src/hooks/useSubject.js

```diff
@@ -0,0 +1,166 @@
+
+    setSubjectList(newSubjectList);
+    localStorage.setItem(KEY, JSON.stringify(newSubjectList));
+  }
+
+  function deleteSubject(state, subjectId) {
+    const newSubjectList = {
+      ...subjectList,
+      [state]: subjectList[state].filter((subject) => subject.id !== subjectId),
+    };
+
+    setSubjectList(newSubjectList);
+    localStorage.setItem(KEY, JSON.stringify(newSubjectList));
+  }
+
+  function addTaskToSubject(state, subjectId, newTaskTitle) {
+    const newTask = {
+      id: uuidv4(),
+      title: newTaskTitle,
+      isCompleted: false,
+    };
+
+    const currentSubjectList = subjectList[state]; // state in which current state is located
+    const currentSubject = currentSubjectList.find(
+      (subject) => subject.id === subjectId
+    );
+
+    const newSubject = {
+      ...currentSubject,
+      taskList: [...currentSubject.taskList, newTask],
+    };
+    const newState = getStateByTaskList(newSubject.taskList);
+
+    toggleSubjectState(state, newState, currentSubjectList, newSubject);
+  }
+
+  function deleteTaskFromSubject(state, subjectId, taskId) {
+    const currentSubjectList = subjectList[state];
+    const currentSubject = currentSubjectList.find(
+      (subject) => subject.id === subjectId
+    );
+
+    const newSubject = {
+      ...currentSubject,
+      taskList: currentSubject.taskList.filter((task) => task.id !== taskId),
+    };
+    const newState = getStateByTaskList(newSubject.taskList);
+
+    toggleSubjectState(state, newState, currentSubjectList, newSubject);
+  }
+
+  function toggleTaskInSubject(state, subjectId, taskId) {
+    const currentSubjectList = subjectList[state];
+    const currentSubject = currentSubjectList.find(
+      (subject) => subject.id === subjectId
+    );
+
+    const newSubject = {
+      ...currentSubject,
+      taskList: currentSubject.taskList.map((task) =>
+        task.id === taskId ? { ...task, isCompleted: !task.isCompleted } : task
+      ),
+    };
+    const newState = getStateByTaskList(newSubject.taskList);
+
+    toggleSubjectState(state, newState, currentSubjectList, newSubject);
+  }
+
+  function toggleSubjectState(
+    state,
+    newState,
+    currentSubjectList,
+    newSubject,
+    subjectId
+  ) {
+    if (state !== newState) {
+      // Case #1: subject is moved to another column
+      const newSubjectList = {
+        ...subjectList,
+        // remove target subject
+        [state]: currentSubjectList.filter(
+          (subject) => subject.id !== newSubject.id
+        ), // add target subject
+        [newState]: [...subjectList[newState], newSubject],
+      };
+
+      setSubjectList(newSubjectList);
+      localStorage.setItem(KEY, JSON.stringify(newSubjectList));
+
+      return;
+    }
+
+    // Case #2: subject remains in same column
+    const newSubjectList = {
+      ...subjectList,
+      [state]: subjectList[state].map((subject) =>
+        subject.id !== subjectId ? { ...subject } : newSubject
+      ),
+    };
```

Comment:
```suggestion
    const newSubjectList = {
      ...subjectList,
      [state]: subjectList[state].map((subject) =>
        subject.id == subjectId ? { ...subject } : newSubject
      ),
    };
```
배포된 링크에서 subject 내에 task 추가가 안되고 있어서 이래저래 디버깅해보다가 위와 같이 바꾸니 제대로 뜨는 것을 확인했습니다! 새로 추가할 task의 subject id가 동일한 해당 subject 내의 태스크를 업데이트 해야한다고 저는 생각해서 수정해보았는데 처음 코드 짜신 의도까진 파악이 안되어서 한 번 더 직접 디버깅 해보시면 좋을 것 같습니다! 😊

Reply:
🤦🤦 사실 목표 subject의 id를 찾아서 변경하는거니 원래 코드대로 `subject.id !== subjectId ? { ...subject } : newSubject` (이것도 `subject.id !== subjectId ? subject : newSubject)`로 쓰면 되긴합니당,,)라고 쓰면 되는데

더 엄청난 문제가 있었네요... 해당 함수의 인자로 `subjectId`를 넘겨주지 않아서 `subjectId` = undefined인 채로 조건문이 돌아 변경이 안됐습니다🤣🤣🤣

꼼꼼히 확인해주셔서 감사합니다😭😭😭😭